compiler/src/dotty/tools/dotc/transform/CheckNoSuperThis.scala

```diff
@@ -0,0 +1,50 @@
+package dotty.tools.dotc
+package transform
+
+import core.*
+import MegaPhase.MiniPhase
+import Contexts.*, Types.*, Symbols.*, SymDenotations.*, Flags.*
+import ast.*
+import Trees.*
+import Decorators.*
+
+import annotation.threadUnsafe
+
+object CheckNoSuperThis:
+  val name: String = "checkNoSuperThis"
+
+/** Checks that super and this calls do not pass `this` as (part of) an argument. */
+class CheckNoSuperThis extends MiniPhase:
+  thisPhase =>
+  import tpd._
+
+  override def phaseName: String = CheckNoSuperThis.name
+
+  override def runsAfterGroupsOf: Set[String] = Set(Constructors.name)
+
+  override def transformDefDef(mdef: DefDef)(using Context): DefDef =
+    if mdef.symbol.isClassConstructor then
+      mdef.rhs match
+        case Block(stats, _) => splitAtSuper(stats) match
```

Comment:
Constructors#transformTemplate also calls splitAtSuper on the constructor, any particular reason the checks here couldn't be done there?

Reply:
We also need to check secondary constructors. It seems to be simpler to do this in a separate miniphase.